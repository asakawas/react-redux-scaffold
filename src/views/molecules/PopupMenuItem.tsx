import React, { ReactElement } from 'react';
import MenuItem, { MenuItemProps } from '@material-ui/core/MenuItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import Typography from '@material-ui/core/Typography';

/**
 * The type of props of PopupMenuItem.
 */
export type PopupMenuItemProps = Readonly<{
  /** The text on the menu item. */
  text: string;

  /** The menu item icon. */
  icon: ReactElement;

  /** The event handler called when the menu item is clicked. */
  onClick: MenuItemProps['onClick'];
}>;

// Wraps with forwardRef so that this component can be given a ref from its parent
// MUI component.
// https://material-ui.com/guides/composition/#caveat-with-refs
const PopupMenuItem = React.forwardRef<HTMLLIElement, PopupMenuItemProps>(
  ({ text, icon, onClick }, ref) => (
    <MenuItem onClick={onClick} ref={ref}>
      <ListItemIcon>{icon}</ListItemIcon>
      <Typography variant="inherit">{text}</Typography>
    </MenuItem>
  ),
);

export default PopupMenuItem;
