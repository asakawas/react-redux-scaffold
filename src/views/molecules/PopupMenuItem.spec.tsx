import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import PopupMenuItem from './PopupMenuItem';

describe('PopupMenuItem', () => {
  test('is rendered as expected', () => {
    const { asFragment } = render(<PopupMenuItem text="test" icon={<></>} onClick={() => {}} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("calls the passed handler when it's clicked", () => {
    const handler = jest.fn();
    const renderResult = render(<PopupMenuItem text="test" icon={<></>} onClick={handler} />);

    const tree = renderResult.getByText('test');
    fireEvent.click(tree);
    expect(handler).toHaveBeenCalledTimes(1);

    fireEvent.click(tree);
    expect(handler).toHaveBeenCalledTimes(2);
  });

  test("contains a node set to 'icon'", () => {
    const renderResult = render(
      <PopupMenuItem text="test" icon={<span>HOGE</span>} onClick={() => {}} />,
    );

    const iconNode = renderResult.getByText('HOGE');
    expect(iconNode).not.toBeNull();
    expect(renderResult.container).toContainElement(iconNode);
  });
});
