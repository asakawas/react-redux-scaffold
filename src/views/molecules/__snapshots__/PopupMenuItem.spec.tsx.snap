// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupMenuItem is rendered as expected 1`] = `
<DocumentFragment>
  <li
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
    role="menuitem"
    tabindex="-1"
  >
    <div
      class="MuiListItemIcon-root"
    />
    <span
      class="MuiTypography-root"
    >
      test
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </li>
</DocumentFragment>
`;
